File: plugins/togetherai-skills/skills/together-evaluations/scripts/run_evaluation.py
#!/usr/bin/env python3
"""
Together AI Evaluations — Run Classify, Score, and Compare (v2 SDK)

Upload an eval dataset, create an evaluation, poll for results, and optionally
download the per-row results file. Supports serverless, dedicated, and external
judge or target models, plus dataset-column evaluation for pre-generated
responses.

Usage:
    python run_evaluation.py --type classify
    python run_evaluation.py --type score --dataset score_prompts.jsonl --eval-column response
    python run_evaluation.py --type compare --model-a-column response_a --model-b-column response_b
    python run_evaluation.py --type classify --eval-model openai/gpt-5 \
        --eval-model-source external --eval-external-api-token "$OPENAI_API_KEY"

Requires:
    uv pip install "together>=2.0.0"
    export TOGETHER_API_KEY=your_key
"""

import argparse
import json
import tempfile
import time
from pathlib import Path
from typing import Any

from together import Together

client = Together()

MODEL_SOURCES = ("serverless", "dedicated", "external")
JUDGE_MODEL = "deepseek-ai/DeepSeek-V4-Pro"
EVAL_MODEL = "Qwen/Qwen3.5-9B"
DEFAULT_EVAL_SYSTEM_TEMPLATE = "You are a helpful assistant."
DEFAULT_INPUT_TEMPLATE = "{{prompt}}"
DEFAULT_CLASSIFY_TEMPLATE = "Classify the following text as positive, negative, or neutral sentiment."
DEFAULT_SCORE_TEMPLATE = (
    "Rate the quality of the response from 1 to 10, where 1 is very poor and 10 is excellent. "
    "Consider accuracy, clarity, and completeness."
)
DEFAULT_COMPARE_TEMPLATE = (
    "Please assess which model has smarter and more helpful responses. Consider clarity, "
    "accuracy, and usefulness."
)


def upload_dataset(dataset: list[dict[str, Any]]) -> str:
    """Write dataset rows to JSONL and upload with purpose=eval."""
    with tempfile.NamedTemporaryFile("w", suffix=".jsonl", delete=False, encoding="utf-8") as temp_file:
        for row in dataset:
            temp_file.write(json.dumps(row) + "\n")
        data_path = Path(temp_file.name)

    try:
        file_response = client.files.upload(file=str(data_path), purpose="eval", check=False)
    finally:
        data_path.unlink(missing_ok=True)

    print(f"Uploaded dataset: {file_response.id}")
    return file_response.id


def load_dataset(path: str | None, fallback_rows: list[dict[str, Any]]) -> list[dict[str, Any]]:
    """Load dataset rows from JSONL, or return bundled sample rows."""
    if not path:
        return fallback_rows

    with open(path, encoding="utf-8") as handle:
        return [json.loads(line) for line in handle if line.strip()]


def poll_evaluation(workflow_id: str, poll_interval: int) -> Any:
    """Poll until the evaluation completes or fails."""
    while True:
        result = client.evals.status(workflow_id)
        print(f"  Status: {result.status}")

        if result.status == "completed":
            return result
        if result.status in ("error", "user_error"):
            print("Evaluation failed")
            return result

        time.sleep(poll_interval)


def result_file_id(result: Any) -> str | None:
    """Return the per-row result file ID when present."""
    results = getattr(result, "results", None)
    if not results:
        return None
    return getattr(results, "result_file_id", None)


def download_result_file(file_id: str, output_path: str) -> None:
    """Download the result JSONL file to a local path."""
    destination = Path(output_path)
    destination.parent.mkdir(parents=True, exist_ok=True)

    with client.files.with_streaming_response.content(id=file_id) as response:
        with open(destination, "wb") as handle:
            for chunk in response.iter_bytes():
                handle.write(chunk)

    print(f"Saved result rows to {destination}")


def build_judge_config(args: argparse.Namespace, default_template: str) -> dict[str, Any]:
    """Build judge model config for serverless, dedicated, or external judges."""
    config: dict[str, Any] = {
        "model": args.judge_model,
        "model_source": args.judge_model_source,
        "system_template": args.judge_system_template or default_template,
    }
    if args.judge_external_api_token:
        config["external_api_token"] = args.judge_external_api_token
    if args.judge_external_base_url:
        config["external_base_url"] = args.judge_external_base_url
    return config


def build_model_config(
    *,
    model: str,
    model_source: str,
    system_template: str,
    input_template: str,
    max_tokens: int,
    temperature: float,
    external_api_token: str | None = None,
    external_base_url: str | None = None,
) -> dict[str, Any]:
    """Build an evaluation target config."""
    config: dict[str, Any] = {
        "model": model,
        "model_source": model_source,
        "system_template": system_template,
        "input_template": input_template,
        "max_tokens": max_tokens,
        "temperature": temperature,
    }
    if external_api_token:
        config["external_api_token"] = external_api_token
    if external_base_url:
        config["external_base_url"] = external_base_url
    return config


def sample_dataset_for_args(args: argparse.Namespace) -> list[dict[str, Any]]:
    """Return a bundled sample dataset that matches the selected workflow."""
    if args.type == "compare" and args.model_a_column and args.model_b_column:
        return [
            {
                "prompt": "Explain the theory of relativity.",
                args.model_a_column: "Relativity explains gravity as the curvature of spacetime.",
                args.model_b_column: "Einstein's theory says mass bends spacetime and changes motion.",
            },
            {
                "prompt": "How does photosynthesis work?",
                args.model_a_column: "Plants convert sunlight, water, and carbon dioxide into sugar.",
                args.model_b_column: "Photosynthesis uses light energy to create glucose and oxygen.",
            },
        ]

    if args.type in {"classify", "score"} and args.eval_column:
        return [
            {
                "prompt": "Summarize what artificial intelligence is.",
                args.eval_column: "Artificial intelligence is software that performs tasks requiring reasoning or prediction.",
            },
            {
                "prompt": "What causes rainbows?",
                args.eval_column: "Rainbows form when water droplets refract, reflect, and disperse sunlight.",
            },
        ]

    samples: dict[str, list[dict[str, Any]]] = {
        "classify": [
            {"prompt": "The product arrived on time and works perfectly!"},
            {"prompt": "Terrible experience. The item was broken."},
            {"prompt": "It's okay, nothing special."},
        ],
        "score": [
            {"prompt": "Explain quantum computing in simple terms."},
            {"prompt": "What causes rainbows?"},
            {"prompt": "How do vaccines work?"},
        ],
        "compare": [
            {"prompt": "Explain the theory of relativity."},
            {"prompt": "What is the meaning of life?"},
            {"prompt": "How does photosynthesis work?"},
        ],
    }
    return samples[args.type]


def maybe_download_results(args: argparse.Namespace, result: Any) -> None:
    """Download result rows when requested and available."""
    file_id = result_file_id(result)
    if file_id:
        print(f"  Result file: {file_id}")
    if args.download_results and file_id:
        download_result_file(file_id, args.download_results)


def run_classify(args: argparse.Namespace, dataset: list[dict[str, Any]]) -> None:
    """Classify evaluation — categorize responses into labels."""
    print("\n=== Classify Evaluation ===")
    file_id = upload_dataset(dataset)

    model_to_evaluate: str | dict[str, Any]
    if args.eval_column:
        model_to_evaluate = args.eval_column
        print(f"Using dataset column for candidate responses: {args.eval_column}")
    else:
        model_to_evaluate = build_model_config(
            model=args.eval_model,
            model_source=args.eval_model_source,
            system_template=args.eval_system_template,
            input_template=args.input_template,
            max_tokens=args.max_tokens,
            temperature=args.temperature,
            external_api_token=args.eval_external_api_token,
            external_base_url=args.eval_external_base_url,
        )

    evaluation = client.evals.create(
        type="classify",
        parameters={
            "input_data_file_path": file_id,
            "judge": build_judge_config(args, DEFAULT_CLASSIFY_TEMPLATE),
            "labels": ["positive", "negative", "neutral"],
            "pass_labels": ["positive"],
            "model_to_evaluate": model_to_evaluate,
        },
    )
    print(f"Created evaluation: {evaluation.workflow_id}")

    result = poll_evaluation(evaluation.workflow_id, poll_interval=args.poll_interval)
    if getattr(result, "results", None):
        print(f"  Label counts: {result.results.label_counts}")
        print(f"  Pass percentage: {result.results.pass_percentage}")
        maybe_download_results(args, result)


def run_score(args: argparse.Namespace, dataset: list[dict[str, Any]]) -> None:
    """Score evaluation — rate responses on a numerical scale."""
    print("\n=== Score Evaluation ===")
    file_id = upload_dataset(dataset)

    model_to_evaluate: str | dict[str, Any]
    if args.eval_column:
        model_to_evaluate = args.eval_column
        print(f"Using dataset column for candidate responses: {args.eval_column}")
    else:
        model_to_evaluate = build_model_config(
            model=args.eval_model,
            model_source=args.eval_model_source,
            system_template=args.eval_system_template,
            input_template=args.input_template,
            max_tokens=args.max_tokens,
            temperature=args.temperature,
            external_api_token=args.eval_external_api_token,
            external_base_url=args.eval_external_base_url,
        )

    evaluation = client.evals.create(
        type="score",
        parameters={
            "input_data_file_path": file_id,
            "judge": build_judge_config(args, DEFAULT_SCORE_TEMPLATE),
            "min_score": 1.0,
            "max_score": 10.0,
            "pass_threshold": 7.0,
            "model_to_evaluate": model_to_evaluate,
        },
    )
    print(f"Created evaluation: {evaluation.workflow_id}")

    result = poll_evaluation(evaluation.workflow_id, poll_interval=args.poll_interval)
    if getattr(result, "results", None):
        scores = result.results.aggregated_scores
        if scores:
            print(f"  Mean score: {scores.mean_score}")
            print(f"  Std score: {scores.std_score}")
            print(f"  Pass percentage: {scores.pass_percentage}")
        maybe_download_results(args, result)


def run_compare(args: argparse.Namespace, dataset: list[dict[str, Any]]) -> None:
    """Compare evaluation — A/B comparison between generated or dataset-column outputs."""
    print("\n=== Compare Evaluation ===")
    file_id = upload_dataset(dataset)

    if args.model_a_column and args.model_b_column:
        model_a: str | dict[str, Any] = args.model_a_column
        model_b: str | dict[str, Any] = args.model_b_column
        print(f"Using dataset columns for comparisons: {args.model_a_column} vs {args.model_b_column}")
    else:
        model_a = build_model_config(
            model=args.model_a,
            model_source=args.model_a_source,
            system_template=args.eval_system_template,
            input_template=args.input_template,
            max_tokens=args.max_tokens,
            temperature=args.temperature,
            external_api_token=args.model_a_external_api_token,
            external_base_url=args.model_a_external_base_url,
        )
        model_b = build_model_config(
            model=args.model_b,
            model_source=args.model_b_source,
            system_template=args.eval_system_template,
            input_template=args.input_template,
            max_tokens=args.max_tokens,
            temperature=args.temperature,
            external_api_token=args.model_b_external_api_token,
            external_base_url=args.model_b_external_base_url,
        )

    parameters: dict[str, Any] = {
        "input_data_file_path": file_id,
        "judge": build_judge_config(args, DEFAULT_COMPARE_TEMPLATE),
        "model_a": model_a,
        "model_b": model_b,
    }
    if args.disable_position_bias_correction:
        parameters["disable_position_bias_correction"] = True
        print("Position-bias correction disabled — running a single judge pass")

    evaluation = client.evals.create(
        type="compare",
        parameters=parameters,
    )
    print(f"Created evaluation: {evaluation.workflow_id}")

    result = poll_evaluation(evaluation.workflow_id, poll_interval=args.poll_interval)
    if getattr(result, "results", None):
        print(f"  A wins: {result.results.a_wins}")
        print(f"  B wins: {result.results.b_wins}")
        print(f"  Ties: {result.results.ties}")
        maybe_download_results(args, result)


def parse_args() -> argparse.Namespace:
    """Parse command-line arguments."""
    parser = argparse.ArgumentParser(description="Together AI evaluations workflow")
    parser.add_argument(
        "--type",
        choices=["classify", "score", "compare"],
        default="classify",
        help="Evaluation workflow to run",
    )
    parser.add_argument("--dataset", help="Path to a JSONL dataset; uses bundled samples when omitted")
    parser.add_argument("--judge-model", default=JUDGE_MODEL, help="Judge model, endpoint ID, or provider shortcut")
    parser.add_argument(
        "--judge-model-source",
        choices=MODEL_SOURCES,
        default="serverless",
        help="Source for the judge model",
    )
    parser.add_argument(
        "--judge-system-template",
        help="Override the default judge Jinja2 template for the selected evaluation type",
    )
    parser.add_argument("--judge-external-api-token", help="API key for an external judge model")
    parser.add_argument("--judge-external-base-url", help="Custom OpenAI-compatible base URL for the judge")
    parser.add_argument("--eval-model", default=EVAL_MODEL, help="Target model for classify or score")
    parser.add_argument(
        "--eval-model-source",
        choices=MODEL_SOURCES,
        default="serverless",
        help="Source for the target model used in classify or score",
    )
    parser.add_argument("--eval-column", help="Dataset column containing pre-generated responses")
    parser.add_argument(
        "--eval-system-template",
        default=DEFAULT_EVAL_SYSTEM_TEMPLATE,
        help="System template for model-based evaluation targets",
    )
    parser.add_argument(
        "--input-template",
        default=DEFAULT_INPUT_TEMPLATE,
        help="Jinja2 input template for model-based evaluation targets",
    )
    parser.add_argument("--max-tokens", type=int, default=512, help="Maximum generation tokens")
    parser.add_argument("--temperature", type=float, default=0.7, help="Generation temperature")
    parser.add_argument("--eval-external-api-token", help="API key for an external evaluation target")
    parser.add_argument("--eval-external-base-url", help="Custom OpenAI-compatible base URL for the target")
    parser.add_argument(
        "--model-a",
        default="Qwen/Qwen3-235B-A22B-Instruct-2507-tput",
        help="Model A for compare evaluations",
    )
    parser.add_argument(
        "--model-a-source",
        choices=MODEL_SOURCES,
        default="serverless",
        help="Source for model A",
    )
    parser.add_argument("--model-a-column", help="Dataset column containing pre-generated model A responses")
    parser.add_argument("--model-a-external-api-token", help="API key for external model A")
    parser.add_argument("--model-a-external-base-url", help="Custom OpenAI-compatible base URL for model A")
    parser.add_argument("--model-b", default=EVAL_MODEL, help="Model B for compare evaluations")
    parser.add_argument(
        "--model-b-source",
        choices=MODEL_SOURCES,
        default="serverless",
        help="Source for model B",
    )
    parser.add_argument("--model-b-column", help="Dataset column containing pre-generated model B responses")
    parser.add_argument("--model-b-external-api-token", help="API key for external model B")
    parser.add_argument("--model-b-external-base-url", help="Custom OpenAI-compatible base URL for model B")
    parser.add_argument(
        "--disable-position-bias-correction",
        action="store_true",
        help=(
            "Compare only: skip the flipped-order judge pass and run a single pass. "
            "Halves judge cost and latency at the expense of position-bias correction."
        ),
    )
    parser.add_argument(
        "--poll-interval",
        type=int,
        default=5,
        help="Seconds between evaluation status checks",
    )
    parser.add_argument(
        "--download-results",
        help="Optional local path for the per-row results JSONL file",
    )
    args = parser.parse_args()

    if (args.model_a_column and not args.model_b_column) or (args.model_b_column and not args.model_a_column):
        parser.error("--model-a-column and --model-b-column must be provided together")
    if args.type != "compare" and (args.model_a_column or args.model_b_column):
        parser.error("--model-a-column and --model-b-column only apply to --type compare")
    if args.type != "compare" and args.disable_position_bias_correction:
        parser.error("--disable-position-bias-correction only applies to --type compare")
    return args


def main() -> None:
    args = parse_args()
    dataset = load_dataset(args.dataset, fallback_rows=sample_dataset_for_args(args))

    if args.type == "classify":
        run_classify(args, dataset)
    elif args.type == "score":
        run_score(args, dataset)
    else:
        run_compare(args, dataset)


if __name__ == "__main__":
    main()

File: plugins/togetherai-skills/skills/together-evaluations/scripts/run_evaluation.ts
#!/usr/bin/env -S npx tsx
/**
 * Together AI Evaluations — Run Classify, Score, and Compare (TypeScript SDK)
 *
 * Upload an eval dataset, create an evaluation, poll for results, and
 * optionally download the per-row results file. Supports serverless,
 * dedicated, and external judge or target models, plus dataset-column
 * evaluation for pre-generated responses.
 *
 * Usage:
 *   npx tsx run_evaluation.ts --type classify
 *   npx tsx run_evaluation.ts --type score --dataset score_prompts.jsonl --eval-column response
 *   npx tsx run_evaluation.ts --type compare --model-a-column response_a --model-b-column response_b
 *   npx tsx run_evaluation.ts --type classify --eval-model openai/gpt-5 \
 *     --eval-model-source external --eval-external-api-token "$OPENAI_API_KEY"
 *
 * Requires:
 *   npm install together-ai
 *   export TOGETHER_API_KEY=your_key
 */

import * as fs from "node:fs";
import * as os from "node:os";
import * as path from "node:path";

import Together from "together-ai";

const client = new Together();

type EvalType = "classify" | "score" | "compare";
type ModelSource = "serverless" | "dedicated" | "external";
type DatasetRow = Record<string, unknown>;

type ScriptArgs = {
  type: EvalType;
  dataset?: string;
  judgeModel: string;
  judgeModelSource: ModelSource;
  judgeSystemTemplate?: string;
  judgeExternalApiToken?: string;
  judgeExternalBaseUrl?: string;
  evalModel: string;
  evalModelSource: ModelSource;
  evalColumn?: string;
  evalSystemTemplate: string;
  inputTemplate: string;
  maxTokens: number;
  temperature: number;
  evalExternalApiToken?: string;
  evalExternalBaseUrl?: string;
  modelA: string;
  modelASource: ModelSource;
  modelAColumn?: string;
  modelAExternalApiToken?: string;
  modelAExternalBaseUrl?: string;
  modelB: string;
  modelBSource: ModelSource;
  modelBColumn?: string;
  modelBExternalApiToken?: string;
  modelBExternalBaseUrl?: string;
  disablePositionBiasCorrection: boolean;
  pollInterval: number;
  downloadResults?: string;
};

const JUDGE_MODEL = "deepseek-ai/DeepSeek-V4-Pro";
const EVAL_MODEL = "Qwen/Qwen3.5-9B";
const DEFAULT_EVAL_SYSTEM_TEMPLATE = "You are a helpful assistant.";
const DEFAULT_INPUT_TEMPLATE = "{{prompt}}";
const DEFAULT_CLASSIFY_TEMPLATE =
  "Classify the following text as positive, negative, or neutral sentiment.";
const DEFAULT_SCORE_TEMPLATE =
  "Rate the quality of the response from 1 to 10, where 1 is very poor and 10 is excellent. Consider accuracy, clarity, and completeness.";
const DEFAULT_COMPARE_TEMPLATE =
  "Please assess which model has smarter and more helpful responses. Consider clarity, accuracy, and usefulness.";

function printHelp(): void {
  console.log(`Together AI evaluations workflow

Flags:
  --type classify|score|compare
  --dataset PATH
  --judge-model MODEL
  --judge-model-source serverless|dedicated|external
  --judge-system-template TEMPLATE
  --judge-external-api-token TOKEN
  --judge-external-base-url URL
  --eval-model MODEL
  --eval-model-source serverless|dedicated|external
  --eval-column COLUMN
  --eval-system-template TEMPLATE
  --input-template TEMPLATE
  --max-tokens N
  --temperature FLOAT
  --eval-external-api-token TOKEN
  --eval-external-base-url URL
  --model-a MODEL
  --model-a-source serverless|dedicated|external
  --model-a-column COLUMN
  --model-a-external-api-token TOKEN
  --model-a-external-base-url URL
  --model-b MODEL
  --model-b-source serverless|dedicated|external
  --model-b-column COLUMN
  --model-b-external-api-token TOKEN
  --model-b-external-base-url URL
  --disable-position-bias-correction
  --poll-interval SECONDS
  --download-results PATH`);
}

const BOOLEAN_FLAGS = new Set(["disable-position-bias-correction"]);

type ParsedFlags = {
  values: Record<string, string>;
  bools: Set<string>;
};

function parseFlagMap(argv: string[]): ParsedFlags {
  const values: Record<string, string> = {};
  const bools = new Set<string>();

  for (let i = 0; i < argv.length; i += 1) {
    const arg = argv[i];
    if (arg === "--help") {
      printHelp();
      process.exit(0);
    }
    if (!arg.startsWith("--")) {
      throw new Error(`Unexpected argument: ${arg}`);
    }
    const flagName = arg.slice(2);
    if (BOOLEAN_FLAGS.has(flagName)) {
      bools.add(flagName);
      continue;
    }
    const next = argv[i + 1];
    if (!next || next.startsWith("--")) {
      throw new Error(`Missing value for ${arg}`);
    }
    values[flagName] = next;
    i += 1;
  }

  return { values, bools };
}

function parseEvalType(value: string | undefined): EvalType {
  const next = value ?? "classify";
  if (next === "classify" || next === "score" || next === "compare") {
    return next;
  }
  throw new Error(`Invalid --type value: ${next}`);
}

function parseModelSource(value: string | undefined, flagName: string): ModelSource {
  const next = value ?? "serverless";
  if (next === "serverless" || next === "dedicated" || next === "external") {
    return next;
  }
  throw new Error(`Invalid ${flagName} value: ${next}`);
}

function parseNumber(value: string | undefined, fallback: number, flagName: string): number {
  if (value === undefined) {
    return fallback;
  }
  const next = Number(value);
  if (!Number.isFinite(next)) {
    throw new Error(`Invalid ${flagName} value: ${value}`);
  }
  return next;
}

function parseScriptArgs(): ScriptArgs {
  const { values: flags, bools } = parseFlagMap(process.argv.slice(2));
  const type = parseEvalType(flags.type);

  const args: ScriptArgs = {
    type,
    dataset: flags.dataset,
    judgeModel: flags["judge-model"] ?? JUDGE_MODEL,
    judgeModelSource: parseModelSource(flags["judge-model-source"], "--judge-model-source"),
    judgeSystemTemplate: flags["judge-system-template"],
    judgeExternalApiToken: flags["judge-external-api-token"],
    judgeExternalBaseUrl: flags["judge-external-base-url"],
    evalModel: flags["eval-model"] ?? EVAL_MODEL,
    evalModelSource: parseModelSource(flags["eval-model-source"], "--eval-model-source"),
    evalColumn: flags["eval-column"],
    evalSystemTemplate: flags["eval-system-template"] ?? DEFAULT_EVAL_SYSTEM_TEMPLATE,
    inputTemplate: flags["input-template"] ?? DEFAULT_INPUT_TEMPLATE,
    maxTokens: parseNumber(flags["max-tokens"], 512, "--max-tokens"),
    temperature: parseNumber(flags.temperature, 0.7, "--temperature"),
    evalExternalApiToken: flags["eval-external-api-token"],
    evalExternalBaseUrl: flags["eval-external-base-url"],
    modelA: flags["model-a"] ?? "Qwen/Qwen3-235B-A22B-Instruct-2507-tput",
    modelASource: parseModelSource(flags["model-a-source"], "--model-a-source"),
    modelAColumn: flags["model-a-column"],
    modelAExternalApiToken: flags["model-a-external-api-token"],
    modelAExternalBaseUrl: flags["model-a-external-base-url"],
    modelB: flags["model-b"] ?? EVAL_MODEL,
    modelBSource: parseModelSource(flags["model-b-source"], "--model-b-source"),
    modelBColumn: flags["model-b-column"],
    modelBExternalApiToken: flags["model-b-external-api-token"],
    modelBExternalBaseUrl: flags["model-b-external-base-url"],
    disablePositionBiasCorrection: bools.has("disable-position-bias-correction"),
    pollInterval: parseNumber(flags["poll-interval"], 5, "--poll-interval"),
    downloadResults: flags["download-results"],
  };

  if ((args.modelAColumn && !args.modelBColumn) || (args.modelBColumn && !args.modelAColumn)) {
    throw new Error("--model-a-column and --model-b-column must be provided together");
  }
  if (args.type !== "compare" && (args.modelAColumn || args.modelBColumn)) {
    throw new Error("--model-a-column and --model-b-column only apply to --type compare");
  }
  if (args.type !== "compare" && args.disablePositionBiasCorrection) {
    throw new Error("--disable-position-bias-correction only applies to --type compare");
  }

  return args;
}

async function uploadDataset(dataset: DatasetRow[]): Promise<string> {
  const dataPath = path.join(os.tmpdir(), `eval_data_${Date.now()}.jsonl`);
  const lines = dataset.map((row) => JSON.stringify(row)).join("\n") + "\n";
  fs.writeFileSync(dataPath, lines, "utf8");

  try {
    const fileResponse = await client.files.upload(dataPath, "eval", false);
    console.log(`Uploaded dataset: ${fileResponse.id}`);
    return fileResponse.id;
  } finally {
    fs.rmSync(dataPath, { force: true });
  }
}

function loadDataset(datasetPath: string | undefined, fallbackRows: DatasetRow[]): DatasetRow[] {
  if (!datasetPath) {
    return fallbackRows;
  }

  return fs
    .readFileSync(datasetPath, "utf8")
    .split("\n")
    .map((line) => line.trim())
    .filter(Boolean)
    .map((line) => JSON.parse(line) as DatasetRow);
}

async function sleep(ms: number): Promise<void> {
  await new Promise((resolve) => setTimeout(resolve, ms));
}

async function pollEvaluation(workflowId: string, pollIntervalSeconds: number): Promise<any> {
  while (true) {
    const result = await client.evals.status(workflowId);
    console.log(`  Status: ${result.status}`);

    if (result.status === "completed") {
      return result;
    }
    if (result.status === "error" || result.status === "user_error") {
      console.error("Evaluation failed");
      return result;
    }

    await sleep(pollIntervalSeconds * 1000);
  }
}

function getResultFileId(result: any): string | undefined {
  return result?.results?.result_file_id;
}

async function downloadResultFile(fileId: string, outputPath: string): Promise<void> {
  fs.mkdirSync(path.dirname(outputPath), { recursive: true });
  const response = await client.files.content(fileId);
  const text = await response.text();
  fs.writeFileSync(outputPath, text, "utf8");
  console.log(`Saved result rows to ${outputPath}`);
}

function buildJudgeConfig(args: ScriptArgs, defaultTemplate: string): Record<string, unknown> {
  const config: Record<string, unknown> = {
    model: args.judgeModel,
    model_source: args.judgeModelSource,
    system_template: args.judgeSystemTemplate ?? defaultTemplate,
  };
  if (args.judgeExternalApiToken) {
    config.external_api_token = args.judgeExternalApiToken;
  }
  if (args.judgeExternalBaseUrl) {
    config.external_base_url = args.judgeExternalBaseUrl;
  }
  return config;
}

function buildModelConfig(options: {
  model: string;
  modelSource: ModelSource;
  systemTemplate: string;
  inputTemplate: string;
  maxTokens: number;
  temperature: number;
  externalApiToken?: string;
  externalBaseUrl?: string;
}): Record<string, unknown> {
  const config: Record<string, unknown> = {
    model: options.model,
    model_source: options.modelSource,
    system_template: options.systemTemplate,
    input_template: options.inputTemplate,
    max_tokens: options.maxTokens,
    temperature: options.temperature,
  };
  if (options.externalApiToken) {
    config.external_api_token = options.externalApiToken;
  }
  if (options.externalBaseUrl) {
    config.external_base_url = options.externalBaseUrl;
  }
  return config;
}

function sampleDatasetForArgs(args: ScriptArgs): DatasetRow[] {
  if (args.type === "compare" && args.modelAColumn && args.modelBColumn) {
    return [
      {
        prompt: "Explain the theory of relativity.",
        [args.modelAColumn]: "Relativity explains gravity as the curvature of spacetime.",
        [args.modelBColumn]: "Einstein's theory says mass bends spacetime and changes motion.",
      },
      {
        prompt: "How does photosynthesis work?",
        [args.modelAColumn]: "Plants convert sunlight, water, and carbon dioxide into sugar.",
        [args.modelBColumn]: "Photosynthesis uses light energy to create glucose and oxygen.",
      },
    ];
  }

  if ((args.type === "classify" || args.type === "score") && args.evalColumn) {
    return [
      {
        prompt: "Summarize what artificial intelligence is.",
        [args.evalColumn]:
          "Artificial intelligence is software that performs tasks requiring reasoning or prediction.",
      },
      {
        prompt: "What causes rainbows?",
        [args.evalColumn]:
          "Rainbows form when water droplets refract, reflect, and disperse sunlight.",
      },
    ];
  }

  if (args.type === "classify") {
    return [
      { prompt: "The product arrived on time and works perfectly!" },
      { prompt: "Terrible experience. The item was broken." },
      { prompt: "It's okay, nothing special." },
    ];
  }
  if (args.type === "score") {
    return [
      { prompt: "Explain quantum computing in simple terms." },
      { prompt: "What causes rainbows?" },
      { prompt: "How do vaccines work?" },
    ];
  }
  return [
    { prompt: "Explain the theory of relativity." },
    { prompt: "What is the meaning of life?" },
    { prompt: "How does photosynthesis work?" },
  ];
}

async function maybeDownloadResults(args: ScriptArgs, result: any): Promise<void> {
  const fileId = getResultFileId(result);
  if (fileId) {
    console.log(`  Result file: ${fileId}`);
  }
  if (args.downloadResults && fileId) {
    await downloadResultFile(fileId, args.downloadResults);
  }
}

async function runClassify(args: ScriptArgs, dataset: DatasetRow[]): Promise<void> {
  console.log("\n=== Classify Evaluation ===");
  const fileId = await uploadDataset(dataset);

  const modelToEvaluate =
    args.evalColumn ??
    buildModelConfig({
      model: args.evalModel,
      modelSource: args.evalModelSource,
      systemTemplate: args.evalSystemTemplate,
      inputTemplate: args.inputTemplate,
      maxTokens: args.maxTokens,
      temperature: args.temperature,
      externalApiToken: args.evalExternalApiToken,
      externalBaseUrl: args.evalExternalBaseUrl,
    });

  if (typeof modelToEvaluate === "string") {
    console.log(`Using dataset column for candidate responses: ${modelToEvaluate}`);
  }

  const evaluation = await client.evals.create({
    type: "classify",
    parameters: {
      input_data_file_path: fileId,
      judge: buildJudgeConfig(args, DEFAULT_CLASSIFY_TEMPLATE),
      labels: ["positive", "negative", "neutral"],
      pass_labels: ["positive"],
      model_to_evaluate: modelToEvaluate,
    },
  });
  console.log(`Created evaluation: ${evaluation.workflow_id}`);

  const result = await pollEvaluation(evaluation.workflow_id!, args.pollInterval);
  if (result.results) {
    console.log(`  Label counts: ${JSON.stringify(result.results.label_counts ?? {})}`);
    console.log(`  Pass percentage: ${result.results.pass_percentage}`);
    await maybeDownloadResults(args, result);
  }
}

async function runScore(args: ScriptArgs, dataset: DatasetRow[]): Promise<void> {
  console.log("\n=== Score Evaluation ===");
  const fileId = await uploadDataset(dataset);

  const modelToEvaluate =
    args.evalColumn ??
    buildModelConfig({
      model: args.evalModel,
      modelSource: args.evalModelSource,
      systemTemplate: args.evalSystemTemplate,
      inputTemplate: args.inputTemplate,
      maxTokens: args.maxTokens,
      temperature: args.temperature,
      externalApiToken: args.evalExternalApiToken,
      externalBaseUrl: args.evalExternalBaseUrl,
    });

  if (typeof modelToEvaluate === "string") {
    console.log(`Using dataset column for candidate responses: ${modelToEvaluate}`);
  }

  const evaluation = await client.evals.create({
    type: "score",
    parameters: {
      input_data_file_path: fileId,
      judge: buildJudgeConfig(args, DEFAULT_SCORE_TEMPLATE),
      min_score: 1.0,
      max_score: 10.0,
      pass_threshold: 7.0,
      model_to_evaluate: modelToEvaluate,
    },
  });
  console.log(`Created evaluation: ${evaluation.workflow_id}`);

  const result = await pollEvaluation(evaluation.workflow_id!, args.pollInterval);
  if (result.results?.aggregated_scores) {
    const scores = result.results.aggregated_scores;
    console.log(`  Mean score: ${scores.mean_score}`);
    console.log(`  Std score: ${scores.std_score}`);
    console.log(`  Pass percentage: ${scores.pass_percentage}`);
    await maybeDownloadResults(args, result);
  }
}

async function runCompare(args: ScriptArgs, dataset: DatasetRow[]): Promise<void> {
  console.log("\n=== Compare Evaluation ===");
  const fileId = await uploadDataset(dataset);

  const modelA =
    args.modelAColumn ??
    buildModelConfig({
      model: args.modelA,
      modelSource: args.modelASource,
      systemTemplate: args.evalSystemTemplate,
      inputTemplate: args.inputTemplate,
      maxTokens: args.maxTokens,
      temperature: args.temperature,
      externalApiToken: args.modelAExternalApiToken,
      externalBaseUrl: args.modelAExternalBaseUrl,
    });
  const modelB =
    args.modelBColumn ??
    buildModelConfig({
      model: args.modelB,
      modelSource: args.modelBSource,
      systemTemplate: args.evalSystemTemplate,
      inputTemplate: args.inputTemplate,
      maxTokens: args.maxTokens,
      temperature: args.temperature,
      externalApiToken: args.modelBExternalApiToken,
      externalBaseUrl: args.modelBExternalBaseUrl,
    });

  if (typeof modelA === "string" && typeof modelB === "string") {
    console.log(`Using dataset columns for comparisons: ${modelA} vs ${modelB}`);
  }

  const parameters: Record<string, unknown> = {
    input_data_file_path: fileId,
    judge: buildJudgeConfig(args, DEFAULT_COMPARE_TEMPLATE),
    model_a: modelA,
    model_b: modelB,
  };
  if (args.disablePositionBiasCorrection) {
    parameters.disable_position_bias_correction = true;
    console.log("Position-bias correction disabled — running a single judge pass");
  }

  const evaluation = await client.evals.create({
    type: "compare",
    parameters: parameters as any,
  });
  console.log(`Created evaluation: ${evaluation.workflow_id}`);

  const result = await pollEvaluation(evaluation.workflow_id!, args.pollInterval);
  if (result.results) {
    console.log(`  A wins: ${result.results.A_wins}`);
    console.log(`  B wins: ${result.results.B_wins}`);
    console.log(`  Ties: ${result.results.Ties}`);
    await maybeDownloadResults(args, result);
  }
}

async function main(): Promise<void> {
  const args = parseScriptArgs();
  const dataset = loadDataset(args.dataset, sampleDatasetForArgs(args));

  if (args.type === "classify") {
    await runClassify(args, dataset);
  } else if (args.type === "score") {
    await runScore(args, dataset);
  } else {
    await runCompare(args, dataset);
  }
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});
